### PR TITLE
Add env variables to windows installer

### LINKED ISSUE #550: Implement relocatable binaries for windows installer
Our OpenSSL build anchors several paths at build time using the OPENSSLDIR/ENGINESDIR/MODULESDIR variables (and subordonate paths based on them),which is unworkable as on windows, with an installer, the selection of the install location is separated from the build time selections.  We need to be able to select those paths at install time and have the openssl binaries read them at run time.  We should use registry keys to store these paths on a per-installation basis.

- [x] Modify the installer to add the appropriate environment variable  keys during install (see below)
- [x] Modify the installer to add the appropriate hklm variables that are read only to installer

Note: This is dependent on #129

### PR DESCRIPTION
**Pull request by nhorman**
add environment variable setting to the installer

this adds a selection screen to the installer, allows the root directory to be set for our various environment variables.  It defaults to the 64 it binary path if included, and falls back to the 32 bit path otherwise, and can be overridden if a user wants to specify a shared directory

### CODE REVIEW

File: windows-installer/openssl.nsi [hunk 1]
```diff
@@ -45,19 +54,19 @@ SectionGroupEnd
...
-# This section is run if installation of the 32 bit binaries are selectd
+# This section is run if installation of the 64 bit binaries are selectd
 SectionGroup "32 Bit Installation"
 	Section "32 Bit Binaries"
 		SetOutPath $INSTDIR\x32\lib
-		File /r "${BUILD32}\Program Files\OpenSSL\lib\"
+		File /r "${BUILD32}\Program Files (x86)\OpenSSL\lib\"
 		SetOutPath $INSTDIR\x32\bin
-		File /r "${BUILD32}\Program Files\OpenSSL\bin\"
+		File /r "${BUILD32}\Program Files(x86)\OpenSSL\bin\"
```
Comment (levitte) [inline windows-installer/openssl.nsi:63]:
```suggestion
		File /r "${BUILD32}\Program Files (x86)\OpenSSL\bin\"
```

**Review by levitte — CHANGES_REQUESTED** [review]
One path error and some indentation nits

File: windows-installer/openssl.nsi [hunk 2]
```diff
@@ -69,30 +78,54 @@ Section "Documentation"
...
+	StrCpy $DataDir "$INSTDIR\x32\Common Files\SSL"
+        StrCpy $ModDir  "$INSTDIR\x32\lib\ossl-modules"
+!endif
+    ifFileExists $INSTDIR\uninstall.exe 0 +2
         ExecWait "$INSTDIR\uninstall.exe /S _?=$INSTDIR"
+
+    WriteRegExpandStr ${openssl_hklm} OPENSSLDIR "$DataDir"
+    WriteRegExpandStr ${openssl_hklm} ENGINESDIR "$ModDir"
+    WriteRegExpandStr ${openssl_hklm} MODULESDIR "$ModDir"
+	SendMessage ${HWND_BROADCAST} ${WM_WININICHANGE} 0 "STR:Environment" /TIMEOUT=5000
```
[the remark this answers is not in the document]
Comment (nhorman) [reply windows-installer/openssl.nsi:116]: yeah, I see some windows newlines as well, will fix

**Review by levitte — APPROVED** [review]
That looks *much* nicer

File: windows-installer/openssl.nsi [hunk 3]
```diff
@@ -1,111 +1,144 @@
...
+# ####################################################
+
+!include "MUI.nsh"
+!include "winmessages.nsh"
+
+!define PRODUCT_NAME "OpenSSL"
+!define VERSION "${MAJOR}.${MINOR}.${PATCH}"
+
+# The name of the output file we create when building this
+# NOTE version is passed with the /D option on the command line
```
Comment (mattcaswell) [inline windows-installer/openssl.nsi:20]: Well, major/minor/patch is passed?

File: windows-installer/openssl.nsi [hunk 4]
```diff
@@ -1,111 +1,144 @@
...
+
+
+Var DataDir
+Var ModDir
+
+Function .onInit
+    StrCpy $INSTDIR "C:\Program Files\openssl-${MAJOR}.${MINOR}"
+FunctionEnd
+
+# This section is run if installation of 32 bit binaries are selected
```
Comment (mattcaswell) [inline windows-installer/openssl.nsi:36]: This comment seems redundant

File: windows-installer/openssl.nsi [hunk 5]
```diff
@@ -1,111 +1,144 @@
...
+    SectionEnd
+    Section "x64 Development Headers"
+        SetOutPath $INSTDIR\x64\include
+        File /r "${BUILD64}\Program Files\OpenSSL\include\"
+    SectionEnd
+SectionGroupEnd
+!endif
+
+!ifdef BUILD32
+# This section is run if installation of the 64 bit binaries are selectd
```
Comment (mattcaswell) [inline windows-installer/openssl.nsi:57]: 32 bit?

File: windows-installer/openssl.nsi [hunk 6]
```diff
@@ -1,111 +1,144 @@
...
+SectionGroup "32 Bit Installation"
+    Section "32 Bit Binaries"
+        SetOutPath $INSTDIR\x32\lib
```
Comment (mattcaswell) [inline windows-installer/openssl.nsi:60]: Should this really be x86 here (and below)? x32 has different connotations for me (https://en.wikipedia.org/wiki/X32_ABI)

File: windows-installer/openssl.nsi [hunk 7]
```diff
@@ -1,111 +1,144 @@
...
+
+!define env_hklm 'HKLM "SYSTEM\CurrentControlSet\Control\Session Manager\Environment"'
+!define openssl_hklm 'HKLM "SOFTWARE\OpenSSL-${MAJOR}.${MINOR}-${CTX}"'
+
+# This is run on uninstall
+Section "Uninstall"
+    RMDIR /r $INSTDIR
+    DeleteRegValue ${openssl_hklm} OPENSSLDIR
+    DeleteRegValue ${openssl_hklm} MODULESDIR
+    DeleteRegValue ${openssl_hklm} ENGINESDIR
```
Comment (mattcaswell) [inline windows-installer/openssl.nsi:94]: I see these reg keys get uninstalled here - but somehow I'm missing the location where they get inserted in the first place?

File: windows-installer/openssl.nsi [hunk 8]
```diff
@@ -1,111 +1,144 @@
...
+!else
+    StrCpy $DataDir "$INSTDIR\x32\Common Files\SSL"
+    StrCpy $ModDir  "$INSTDIR\x32\lib\ossl-modules"
+!endif
+    ifFileExists $INSTDIR\uninstall.exe 0 +2
+        ExecWait "$INSTDIR\uninstall.exe /S _?=$INSTDIR"
+
+    WriteRegExpandStr ${openssl_hklm} OPENSSLDIR "$DataDir"
+    WriteRegExpandStr ${openssl_hklm} ENGINESDIR "$ModDir"
+    WriteRegExpandStr ${openssl_hklm} MODULESDIR "$ModDir"
```
Comment (mattcaswell) [inline windows-installer/openssl.nsi:115]: This looks like we are inserting these keys on uninstall which I don't understand. Can you explain what is happening here?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: windows-installer/openssl.nsi (hunk 8 above)
Comment (nhorman) [reply windows-installer/openssl.nsi:115]: CheckRunUninstaller is run universally on exit from the DIRECTORYPAGE_TEXT_DESTINATION panel.  It actually only runs the uninstaller if it exists. see:
```
ifFileExists $INSTDIR\uninstall.exe 0 +2
        ExecWait "$INSTDIR\uninstall.exe /S _?=$INSTDIR"
```

But I agree, its a bit fuzzy.  I can refactor that

File: windows-installer/openssl.nsi (hunk 7 above)
Comment (nhorman) [reply windows-installer/openssl.nsi:94]: Its in CheckRunUninstaller

**Review by mattcaswell — APPROVED** [review]

**Comment by nhorman** [comment]
merged, thank you!

**Pull request closed without merging**